swarm/storage/pushsync/pusher.go

```diff
@@ -0,0 +1,277 @@
+		pushed:   make(map[string]*pushedItem),
+		receipts: make(chan chunk.Address),
+		ps:       ps,
+	}
+	go p.sync()
+	return p
+}
+
+// Close closes the pusher
+func (p *Pusher) Close() {
+	close(p.quit)
+}
+
+// sync starts a forever loop that pushes chunks to their neighbourhood
+// and receives receipts (statements of custody) for them.
+// chunks that are not acknowledged with a receipt are retried
+// not earlier than retryInterval after they were last pushed
+// the routine also updates counts of states on a tag in order
+// to monitor the proportion of saved, sent and synced chunks of
+// a file or collection
+func (p *Pusher) sync() {
+	var chunks <-chan chunk.Chunk
+	var cancel, stop func()
+	var ctx context.Context
+	var synced []storage.Address
+
+	// timer
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	// register handler for pssReceiptTopic on pss pubsub
+	deregister := p.ps.Register(pssReceiptTopic, false, func(msg []byte, _ *p2p.Peer) error {
+		return p.handleReceiptMsg(msg)
+	})
+	defer deregister()
+
+	chunksInBatch := -1
+	var batchStartTime time.Time
+
+	for {
+		select {
+
+		// handle incoming chunks
+		case ch, more := <-chunks:
+			chunksInBatch++
+			// if no more, set to nil and wait for timer
+			if !more {
+				chunks = nil
+				continue
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk", nil).Inc(1)
+			// if no need to sync this chunk then continue
+			if !p.needToSync(ch) {
+				continue
+			}
+
+			metrics.GetOrRegisterCounter("pusher.send-chunk.send-to-sync", nil).Inc(1)
+			// send the chunk and ignore the error
+			if err := p.sendChunkMsg(ch); err != nil {
+				log.Error("error sending chunk", "addr", ch.Address(), "err", err)
+			}
+
+			// handle incoming receipts
+		case addr := <-p.receipts:
+			metrics.GetOrRegisterCounter("pusher.receipts.all", nil).Inc(1)
+			log.Debug("synced", "addr", addr)
+			// ignore if already received receipt
+			item, found := p.pushed[addr.Hex()]
+			if !found {
+				metrics.GetOrRegisterCounter("pusher.receipts.not-found", nil).Inc(1)
+				log.Debug("not wanted or already got... ignore", "addr", addr)
+				continue
+			}
+			if item.synced {
+				metrics.GetOrRegisterCounter("pusher.receipts.already-synced", nil).Inc(1)
+				log.Debug("just synced... ignore", "addr", addr)
+				continue
+			}
+			metrics.GetOrRegisterCounter("pusher.receipts.synced", nil).Inc(1)
+			// collect synced addresses
+			synced = append(synced, addr)
+			// set synced flag
+			item.synced = true
+			// increment synced count for the tag if exists
+			if item.tag != nil {
+				item.tag.Inc(chunk.StateSynced)
+			}
+
+			// retry interval timer triggers starting from new
+		case <-timer.C:
+			metrics.GetOrRegisterCounter("pusher.subscribe-push", nil).Inc(1)
+			// TODO: implement some smart retry strategy relying on sent/synced ratio change
+			// if subscribe was running, stop it
+			if stop != nil {
+				stop()
+			}
+			for _, addr := range synced {
```

Comment:
Why are we not removing chunks right in the handle receipt? Why do we have to wait for the timer to kick in again to clear our cache of synced chunks?

Reply:
in order to allow much more efficient batch set in localstore

Reply:
This seems rather weird. We are not processing chunks efficiently in go-routines, but we have added data structures to maintain state in order to use the LocalStore efficiently?

Have we measured if there is any noticeable difference, so that we must add this cache?

Reply:
If fail to see what you mean by `efficient batch set` - 

```
			for _, addr := range synced {
				// set chunk status to synced, insert to db GC index
				if err := p.store.Set(context.Background(), chunk.ModeSetSync, addr); err != nil {
					log.Warn("error setting chunk to synced", "addr", addr, "err", err)
					continue
				}
				delete(p.pushed, addr.Hex())
			}
```

We might as well just call the `Set` when the chunk is synced and avoid having this cache.

Reply:
now using multi set